nodes/student_localiser.py

```diff
@@ -0,0 +1,660 @@
+					self.make_new_goal()
+					return
+			else:
+				path_correction = 1.0
+				path_correction_distance = 0.0
+
+			self.sum_theta_correction += rotation_correction
+			self.sum_path_correction += path_correction_distance
+
+			goal_offset = get_corrected_goal_offset(current_pose_odom, next_goal_odom, rotation_correction, path_correction)
+			new_goal = current_pose_odom * goal_offset
+			self.update_goal(new_goal, False, turning_goal)
+
+	def do_discrete_correction(self, pose, old_goal_frame_odom, old_goal_frame_world):
+		new_goal_frame_world = tf_conversions.fromMsg(self.poses[self.goal_index])
+		turning_goal = is_turning_goal(old_goal_frame_world, new_goal_frame_world)
+		inter_goal_offset_world = old_goal_frame_world.Inverse() * new_goal_frame_world
+		inter_goal_distance_world = inter_goal_offset_world.p.Norm()
+
+		offsets, correlations = self.calculate_image_pose_offset(self.goal_index, self.search_range)
+		if self.goal_index >= self.search_range:
+			rotation_offset = offsets[self.search_range]
+			rotation_correlation = correlations[self.search_range]
+		else:
+			rotation_offset = offsets[-self.search_range-1]
+			rotation_correlation = correlations[-self.search_range-1]
+
+		offset = rotation_offset
+
+		rotation_correction = self.rotation_correction_gain * offset
+		if rotation_correlation < self.image_recognition_threshold:
+			rotation_correction = 0.0
+
+		if not turning_goal and self.goal_index >= self.search_range and self.goal_index < len(self.poses)-self.search_range:
+			corr = np.array(correlations)
+			corr -= self.image_recognition_threshold
+			corr[corr < 0] = 0.0
+			s = corr.sum()
+			if s > 0:
+				corr /= s
+			w = corr * np.arange(-self.search_range,self.search_range+1,1)
+			pos = w.sum()
+			path_error = pos
+
+			path_correction_distance = -self.path_correction_gain * path_error * GOAL_DISTANCE_SPACING
+			path_correction = (GOAL_DISTANCE_SPACING + path_correction_distance) / GOAL_DISTANCE_SPACING
+			if np.isnan(path_correction):
+				print(corr, s, w, pos, GOAL_DISTANCE_SPACING)
+			if -path_correction_distance > GOAL_DISTANCE_SPACING:
+				print('PATH CORRECTION ERROR: correction is greater than distance to goal!')
+				print('corr = %s; pos = %f, path_correction = %f, goal_distance = %f' % (str(corr),pos,path_correction_distance,GOAL_DISTANCE_SPACING))
+				print('path_correction = %f' % path_correction)
+				path_correction_distance = -GOAL_DISTANCE_SPACING
+				path_correction = 0.0
+		else:
+			path_correction = 1.0
+			path_correction_distance = 0.0
+
+		self.sum_theta_correction = rotation_correction
+		self.sum_path_correction = path_correction_distance
+		return rotation_correction, path_correction
+
+	def calculate_image_pose_offset(self, image_to_search_index, half_search_range=0):
+		if self.last_image is not None:
+			match_request = ImageMatchRequest(image_processing.image_to_msg(self.last_image), UInt32(image_to_search_index), UInt32(half_search_range))
```

Comment:
Then the localiser gets stuck waiting here for the image matcher to finally return? Do you get any problems with this call timing out or it seems to work fine?

Reply:
The localiser also have the number of frames delayed in one of the function. So there was not a problem with timing out.
For example, if the frames_dealy is set as 10, the second robot will be waiting 10 frames behind the teacher's robot until it receives another pose/image data.

Reply:
Ok, I'll leave some more questions there